Article cdn fastly cloudfront

mezis opened the pull request:
First post!

cc @jphastings, our beloved Editor-in-Chief. Feel free to rewrite any part of this.

To read the article locally:

- Add `future: true` to `_config.yml`
- Run `bundle exec jekyll serve`

sdorunga1 commented on _posts/2016-07-13-split-testing-cdns.md:
```diff
@@ -0,0 +1,352 @@
+TL,DR:
+
+- Fastly is generally more performant than AWS Cloudfront — for our userbase.
+- A/B testing CDNs is easy (with out toolset).
```
with out => without/with our?

jphastings replied:
`s/out/our/`?

sdorunga1 commented on _posts/2016-07-13-split-testing-cdns.md:
```diff
@@ -0,0 +1,352 @@
+
+## Context
+
+We currently use a CDN[^cdn]
```
This doesn't seem to render right? Or is it just the markdown preview on github acting funny?

tommyp replied:
I think it's the latter.

sdorunga1 commented on _posts/2016-07-13-split-testing-cdns.md:
```diff
@@ -0,0 +1,352 @@
+
+Our current HTTP infrastructure has 3 tiers:
+
+- Client requests hit either a CDN (assets) or our reverse proxy (NGinx).
```
*Nginx

sdorunga1 commented on _posts/2016-07-13-split-testing-cdns.md:
```diff
@@ -0,0 +1,352 @@
+Both the proxy and the Heroku app are hosted by AWS, in the `eu-west-1` (Dublin)
+datacenter.
+
+## Testable instructure variants
```
*infrastructure

sdorunga1 commented on _posts/2016-07-13-split-testing-cdns.md:
```diff
@@ -0,0 +1,352 @@
+
+
+![Infrastructure variants](/images/2016-07-13-split-testing-cdns--03.png)
+{: .dg-figure}
```
Same as the footnote above and the one below

mezis replied:
Github Pages uses [Kramdown](https://help.github.com/articles/updating-your-markdown-processor-to-kramdown/) — the renderer in PRs uses [GFM](https://help.github.com/articles/about-writing-and-formatting-on-github/).

Welcome to the world of concurrent Markdown implementations.

sdorunga1 commented on _posts/2016-07-13-split-testing-cdns.md:
```diff
@@ -0,0 +1,352 @@
+all synchronous assets have been laoded and the page has been rendered).
+
+This is the part of the page load lifecycle where differences in asset load
+timings may have an impact. Given all other things are unchanges, it's the
```
*unchanged

jphastings commented on _posts/2016-07-13-split-testing-cdns.md:
```diff
@@ -0,0 +1,352 @@
+
+---
+
+TL,DR:
```
Having a TL,DR; is great, but it'd be great to frame the article with a short paragraph first. Call me buzz-feed generation, but it'd be good to build some excitement about the article here & give folks a reason to read it, instead of a reason to not read it 😛

mezis replied:
The exerpt (in the Yaml front matter) will be visible both in the list of posts on on the post page — is that what you mean?

I can make that intro more intro-y if you want though.

sdorunga1 commented on _posts/2016-07-13-split-testing-cdns.md:
```diff
@@ -0,0 +1,352 @@
+
+- [Amazon Cloudfront][cloudfront] and [Fastly][fastly]: the two CDNs we've
+  pitted against eachother.
+- New Relic [Synthetics][nrsynth]: a Pindgom clone that lest up repeatedly call
```
lest up => let's us?

sdorunga1 commented on _posts/2016-07-13-split-testing-cdns.md:
```diff
@@ -0,0 +1,352 @@
+- [Amazon Cloudfront][cloudfront] and [Fastly][fastly]: the two CDNs we've
+  pitted against eachother.
+- New Relic [Synthetics][nrsynth]: a Pindgom clone that lest up repeatedly call
+  and HTTP endpoint and get statistics on response time.
```
and => an

tommyp commented on _posts/2016-07-13-split-testing-cdns.md:
```diff
@@ -0,0 +1,352 @@
+page and reports on performance — somewhat like a dumbed-down
+[Webpagetest](http://www.webpagetest.org/), but on every single page view.
+
+The JS library lets us "tag" each pageview with a metadata that we can later
```
`a metadata`?

jphastings commented on _posts/2016-07-13-split-testing-cdns.md:
```diff
@@ -0,0 +1,352 @@
+- Synthetic performance checks (e.g. with Pingdom) do not reflect end user
+  performance experience.
+
+We split tested Cloudfront vs. Fastly across 1,000,000 users, in 12
```
This sentence is an impressive statement to make; I'm not sure if Jekyll give us easy-to-use pull-quotes but I'd love to make this more prominent; it might be a bit too cheesy, but we could embolden the numbers out so they stand out. What do you think of something like this?

> We split our production traffic between Cloudfront and Fastly's CDN networks, looking for significant (_p_ < 0.01) differences in page load times and conversion across **1,000,000 users** in **12 countries** on **3 continents**. We found Fastly a massive **50% faster** for page load times, which translated into a whopping **3.9% increase in conversion**.
>
> Naturally there are some important caveats to these statistic, and we think the processes we've used to run this experiment are interesting enough to blog about, so read on for the detail…

jphastings replied:
The above does accentuate the page load -> conversion link which @sdorunga1 has questions about, it may not be an appropriate link to make!

sdorunga1 commented on _posts/2016-07-13-split-testing-cdns.md:
```diff
@@ -0,0 +1,352 @@
+We split tested Cloudfront vs. Fastly across 1,000,000 users, in 12
+countries, on 3 continents, over a significant time period. Using Fastly
+improved page load times by up to 50% and site conversion by up to a whopping
+3.9% (with _p_ < 0.01). Read on for the caveats!
```
Those are some pretty serious caveats though 😸 It is quite a bit misleading though, I think it sets up very high expectations compared to the outcome in the end. Which is basically there's a marginal increase in performance outside of AU and HK and the conversion confidence intervals are quite all over the place which from my entirely vague understanding of statisticsI wouldn't count as statistically significant increases in conversion apart from DE and slightly for FR which is also a bit suspicious because they are not among the top performance increases. I'm just pointing it out because the abstract seems to link conversion to performance which is not what you're claiming throughout the article. You may have done this on purpose but that's my 2 cens

mezis replied:
Actually, the performance improvement is pretty massive locally, and very good globally.
It's on _conversion_ that we don't see the expected improvements.

I'll try to rephrase a bit, see if I can convince you.

sdorunga1 replied:
Yep, that's basically what I meant. The way the text was worded it did make it sound like a very tight and quite high relationship existed between performance and conversion (which is what is currently believed to generally be true) which would form a very different expectation in the mind of a reader.

I'm very happy with the new wording as it's making it a lot clearer what the outcome actually was, why it's significant/useful and how you came up with those conclusions which is really valuable.

jphastings commented on _posts/2016-07-13-split-testing-cdns.md:
```diff
@@ -0,0 +1,352 @@
+
+We set up 4 variants of our infrastrucutre:
+`cdn0` is the existing, Cloudfront-based setup - the control.
+`cdn1` to `cdn3` to replaces Cloudfront with various Fastly configurations:
```
`s/to replaces/replaced/`

jphastings commented on _posts/2016-07-13-split-testing-cdns.md:
```diff
@@ -0,0 +1,352 @@
+`cdn0` is the existing, Cloudfront-based setup - the control.
+`cdn1` to `cdn3` to replaces Cloudfront with various Fastly configurations:
+
+- `cdn1` uses Fastly's "shield" feature, whereby local POPs[^pop] are backed by a
```
Link to product page for the shield feature?

jphastings commented on _posts/2016-07-13-split-testing-cdns.md:
```diff
@@ -0,0 +1,352 @@
+
+---
+
+## Context
```
Something more engaging?

jphastings commented on _posts/2016-07-13-split-testing-cdns.md:
```diff
@@ -0,0 +1,352 @@
+### Synthetic performance comparison
+
+To get a sense of what performance differences we could observe between
+countries, we set up a synthetic check per test variant in NR Synthetics.
```
`[NR Synthetics](#tools-used)`?

jphastings commented on _posts/2016-07-13-split-testing-cdns.md:
```diff
@@ -0,0 +1,352 @@
+countries, we set up a synthetic check per test variant in NR Synthetics.
+
+Each of these checks is configured to request a given asset (our main CSS file)
+every minute, from a number of different locations across the globe; NR Insights
```
`[NR Insights](#tools-used)`?

jphastings commented on _posts/2016-07-13-split-testing-cdns.md:
```diff
@@ -0,0 +1,352 @@
+  the CDN's upstream. This most closely mimics our Cloudfront setup.
+- `cdn3` is backed directly by the Heroku application.
+
+Each of those is given a different DNS entry for each of our TLDs,
```
You've swapped tenses here - "we **set up** 4 variants" vs "Each of those **is** given" - I think the past tense was working for you, could we stick with that?

jphastings commented on _posts/2016-07-13-split-testing-cdns.md:
```diff
@@ -0,0 +1,352 @@
+To get a sense of what performance differences we could observe between
+countries, we set up a synthetic check per test variant in NR Synthetics.
+
+Each of these checks is configured to request a given asset (our main CSS file)
```
If we're going with past tense this `is` should be a `was`.

jphastings commented on _posts/2016-07-13-split-testing-cdns.md:
```diff
@@ -0,0 +1,352 @@
+every minute, from a number of different locations across the globe; NR Insights
+lets us access detailed reports of performance for each individual request.
+
+Because Cloudfront is our existing setup, We allowed for a 1-week warmup period:
```
`s/We/we/`

jphastings commented on _posts/2016-07-13-split-testing-cdns.md:
```diff
@@ -0,0 +1,352 @@
+![Infrastructure variants](/images/2016-07-13-split-testing-cdns--03.png)
+{: .dg-figure}
+
+At first glance, this looks like Cloudfront (`cdn0`) is massively winning over
```
"this" is slightly unspecified, can we go with a clearer subject, eg. "At first glance this graph seems to show that Cloudfront (`cdn0`) …"?

jphastings commented on _posts/2016-07-13-split-testing-cdns.md:
```diff
@@ -0,0 +1,352 @@
+At first glance, this looks like Cloudfront (`cdn0`) is massively winning over
+Fastly (`cdn[1-3]`). The trick here is that Cloudfront is part of Amazon's
+infrastructure, so it's unfairly advantaged whenever the pinging server is also
+in AWS — most of the cases here.
```
I'm not quite sure what "most of the cases here" is trying to say - can we expand it a little bit?

jphastings commented on _posts/2016-07-13-split-testing-cdns.md:
````diff
@@ -0,0 +1,352 @@
+```
+
+From thereon, it's just a matter of waiting enough time for enough data to be
+harvested and reaching statistical significance.
````
"for enough data to be harvested_to_ reaching statistical significance."?

jphastings commented on _posts/2016-07-13-split-testing-cdns.md:
```diff
@@ -0,0 +1,352 @@
+> the user-agent header for each HTTP request.
+{: .dg-sidebar}
+
+For this second test, we'll run a proper A/B test trying to falsify the
```
`we'll run` is future tense - we should probably stick to the past as above?

jphastings commented on _posts/2016-07-13-split-testing-cdns.md:
```diff
@@ -0,0 +1,352 @@
+
+### Performance for actual users
+
+> ##### Why split the analysis by country _and_ platform?
```
Might be worth moving this section below the next paragraph, as on mobile these sections will appear in-line (#12) and the the main paragraph introduces the decision to do the thing this one is explaining :)

mezis replied:
Looks terrible on desktop - but point taken. Will move it.

jphastings replied:
Hmmm. May be worth leaving until after we know how we're going to tackle #12 then!

jphastings commented on _posts/2016-07-13-split-testing-cdns.md:
```diff
@@ -0,0 +1,352 @@
+
+This is the part of the page load lifecycle where differences in asset load
+timings may have an impact. Given all other things are unchanges, it's the
+_only_ factor of change, which is what we're after in an A/B test.
```
These seems a little tautological: if everything else is unchanged then this is the only thing that changed. I'm not quite sure what point you were trying to make beyond that experiments should be limited to only one dynamic variable (an excellent point to make!) so I'm not sure how I'd rephrase this - any ideas?

jphastings commented on _posts/2016-07-13-split-testing-cdns.md:
````diff
@@ -0,0 +1,352 @@
+LIMIT 20
+```
+
+That's not extremely convenient, and we'll need to post-process anyhow (to run
````
"The HTML output isn't very convenient but we needed to post-process (to run …), so …" makes the subject of this sentence much clearer ("That" probably isn't the right determiner here, you'd want to use "this", but an explicit subject is nice as the reader has just been deep in code 😊)

jphastings commented on _posts/2016-07-13-split-testing-cdns.md:
```diff
@@ -0,0 +1,352 @@
+hour or copying-and-pasting into Excel follows.
+
+Intermission: for proper A/B testing, have a read of [Evan Miller's Awesome
+A/B Testing Tools](http://www.evanmiller.org/ab-testing/). One of my favourite
```
:+1:

jphastings commented on _posts/2016-07-13-split-testing-cdns.md:
```diff
@@ -0,0 +1,352 @@
+{: .dg-figure}
+
+This graphs the ratio of load+render times for Fastly (`cdn1` variant, the best
+one) versus Cloudfront (`cdn0`), split by the users' countries - anything on the
```
"(`cdn0` variant, our pre-existing set up)" for balance?

jphastings commented on _posts/2016-07-13-split-testing-cdns.md:
```diff
@@ -0,0 +1,352 @@
+right, Fastly wins.
+
+The error bars are for _p < 0.05_, not good enough for Science or Nature but
+generally accepted as good enough for multivariate tests in tech companies.
```
If there's a reasonable citation you can offer here it'd be really interesting, though I don't think it's needed for the article :)

jphastings commented on _posts/2016-07-13-split-testing-cdns.md:
```diff
@@ -0,0 +1,352 @@
+Unlike for our earlier synthetic tests, the results are quite clear cut:
+
+- Fastly is significantly faster almost everywhere.
+- Notable exceptions exist: the UAE (Fastly doesn't ahve a POP there yet); Spain
```
`s/ahve/have/`

jphastings commented on _posts/2016-07-13-split-testing-cdns.md:
```diff
@@ -0,0 +1,352 @@
+
+- Fastly is significantly faster almost everywhere.
+- Notable exceptions exist: the UAE (Fastly doesn't ahve a POP there yet); Spain
+  on desktops (no obvious explanation), Ireland and Singapore on mobiles (Fastly
```
This was a semicolon delimited list a moment ago - can we `s/,/;/`? 😊

jphastings commented on _posts/2016-07-13-split-testing-cdns.md:
```diff
@@ -0,0 +1,352 @@
+- Notable exceptions exist: the UAE (Fastly doesn't ahve a POP there yet); Spain
+  on desktops (no obvious explanation), Ireland and Singapore on mobiles (Fastly
+  probably has a weaker presence there than Cloudfront — likely as the capitals
+  there are Amazon datacenter cities).
```
I'd be tempted to break these exceptions out into their own sub-bullet points so you can address them each more clearly (rather than in parentheses) - this last one is particularly interesting!

jphastings commented on _posts/2016-07-13-split-testing-cdns.md:
```diff
@@ -0,0 +1,352 @@
+  there are Amazon datacenter cities).
+
+At this point we could opt to jsut switch over to Fastly and bite the bullet in
+places where it's slightly weaker; but what ultimately matters is not how fast
```
Italics for emphasis? I'm in two minds…

> but what ultimately matters is not _how fast_ our site and apps are, but _how many users we satisfy_.

jphastings commented on _posts/2016-07-13-split-testing-cdns.md:
```diff
@@ -0,0 +1,352 @@
+
+### Impact on conversion
+
+For this last test, we're using the same A/B test setup as above, but this time
```
Tense here again; I don't mind swapping to present for the article as you seem to favour it in the latter parts, but a consistent one would be good :)

jphastings commented on _posts/2016-07-13-split-testing-cdns.md:
```diff
@@ -0,0 +1,352 @@
+we're measuring conversion across CDN variants, countries, and platforms.
+
+We use [Segment][segment] to relay and store our analytics. The gist is that
+form anywhere in our backend, we can say:
```
> The gist is that form anywhere in our backend, we can say:
This sounds a little strange, maybe something like the following?
> It allows us to track events, including custom attributes, from anywhere in our back end application:

Also, if you aim to keep the original `s/form/from/`

jphastings commented on _posts/2016-07-13-split-testing-cdns.md:
````diff
@@ -0,0 +1,352 @@
+$segment.track 'Completed Order', cdn: $cdn
+```
+
+and end up with a record in Redshift, tagged with the user who made the request,
````
I'm not such a fan of code blocks being part of a sentence like this; it doesn't flow so well for people unfamiliar with the code you're presenting. You could restart the text with something like "This inserts a record in Redshift…"

jphastings commented on _posts/2016-07-13-split-testing-cdns.md:
````diff
@@ -0,0 +1,352 @@
+```
+
+and end up with a record in Redshift, tagged with the user who made the request,
+the browser, the country, the CDN variant used, and a whole lot of other bits of
````
Parenthesise the last point for clarity?

> the country and the CDN variant used (as well as a whole lot of other bits of metadata we won't use here.)

jphastings commented on _posts/2016-07-13-split-testing-cdns.md:
```diff
@@ -0,0 +1,352 @@
+the browser, the country, the CDN variant used, and a whole lot of other bits of
+metadata we won't use here.
+
+Again, once that's set up, it's just a matter of waiting until there's enough
```
it's is present tense, if you're changing to past :)

jphastings commented on _posts/2016-07-13-split-testing-cdns.md:
```diff
@@ -0,0 +1,352 @@
+conversion change (+0.0% ±0.8%, _p_ = 0.5), or a slight improvement (+0.9%
+±0.9%, _p_ = 0.02).
+
+But once split by country and platform results are surprising.
```
We need a break in there to aid reading, something like this? 

> But once split by country and platform the results are surprising.

jphastings commented on _posts/2016-07-13-split-testing-cdns.md:
```diff
@@ -0,0 +1,352 @@
+But once split by country and platform results are surprising.
+
+
+![Infrastructure variants](/images/2016-07-13-split-testing-cdns--05.png)
```
Title of graph has `s/when/with/` I believe

jphastings commented on _posts/2016-07-13-split-testing-cdns.md:
```diff
@@ -0,0 +1,352 @@
+  conversion.
+- Most other countries have conversion and performance trending in the same
+  direction.
+- But... while Australian mobiles see a significant improvement in performance
```
`while … but` sadly doesn't scan; we can replace `s/ but/,/` or `s/while //` (we can't use both forms at once).

jphastings commented on _posts/2016-07-13-split-testing-cdns.md:
```diff
@@ -0,0 +1,352 @@
+  but conversion is significantly worse!
+
+We don't have an explanation for this last point. It's possible that major
+Aussie ISPs do something bad to our HTTP requests that cuase some to fail, but
```
`s/cuase/cause/`

jphastings commented on _posts/2016-07-13-split-testing-cdns.md:
```diff
@@ -0,0 +1,352 @@
+
+We don't have an explanation for this last point. It's possible that major
+Aussie ISPs do something bad to our HTTP requests that cuase some to fail, but
+we haven't been able to reproduce error cases or received error reports form our
```
Generally an 'or' split like this implies the use of the same verb before the latter part, which implies "we haven't been able to received error reports", making this sound a little odd. You could add the pronoun again to make it clearer: "able to reproduce error cases nor have we received error".

jphastings commented on _posts/2016-07-13-split-testing-cdns.md:
```diff
@@ -0,0 +1,352 @@
+we haven't been able to reproduce error cases or received error reports form our
+local teams or local customers.
+
+This leave us with our heads scratching, and another example that AB tests
```
`s/leave/leaves/`

jphastings commented on _posts/2016-07-13-split-testing-cdns.md:
```diff
@@ -0,0 +1,352 @@
+aren't perfect: they can give strong results... but they can also give very
+inconsistent results if there's a hidden variable.
+
+Our take on this is to, as we did above, couple A/B testing with other
```
I'd stick the subclause after the content of the sentence it's hard for the reader to reference what we did above when they don't know the context of the sentence is yet:

> Our take on this was to couple A/B testing with other sources that we know correlate to conversion, as we did above.

mezis replied:
I'm rewriting this section in part so - maybe have another read after?

jphastings commented on _posts/2016-07-13-split-testing-cdns.md:
```diff
@@ -0,0 +1,352 @@
+## Tools used
+
+- [Amazon Cloudfront][cloudfront] and [Fastly][fastly]: the two CDNs we've
+  pitted against eachother.
```
`s/eachother/each other/` for the pedantic amongst… me 😛

jphastings commented:
Really interesting! Thanks Julien, it's a great article. I've made a couple of notes here and there with one general point around going with a single tense for the article (past or present I'd recommend - though bonus points for something wacky 😛)

mezis commented:
@jphastings, I think I've addressed most of your comments.
I've also revamped the intro and conclusion, and added a couple graphs to address @sdorunga1's valid concerns.

sdorunga1 commented:
Very good articl! I think showing exactly how we came up with the numbers and the very detailed explanations of what you chose to ignore/take into account were very clear and useful. Especially compared with just having a load of data presented as graphs and a vague conclusion which is what most articles do :)

Are we still running the AB test btw?

mezis commented:
@sdorunga1: Thanks! Yeah, we're still running the test. I'll probably leave it up until we're ready to go properly live with Fastly (sorting out proper SSL certs for now), because, why not get more data and update the graphs.

@jphastings: as I'm on hols next week, can I leave this PR with you? You're welcome to change anything. Seems to me it's almost good to merge.

jphastings commented on _posts/2016-07-13-split-testing-cdns.md:
```diff
@@ -0,0 +1,408 @@
+Our current HTTP infrastructure has 3 tiers:
+
+- Client requests hit either a CDN (assets) or our reverse proxy
+  ([Nginx][nginx]).
```
No longer links anywhere :(

jphastings commented on _posts/2016-07-13-split-testing-cdns.md:
```diff
@@ -0,0 +1,408 @@
+- Client requests hit either a CDN (assets) or our reverse proxy
+  ([Nginx][nginx]).
+- The CDN loads from the proxy on misses.
+- The proxy is backed by a [Heroku][heroku]-hosted application.
```
Nor this one :()

jphastings commented:
@mezis I left it until you were back in the office (welcome back! 🏖) but I think we're almost ready to publish - some mini issues I've commented on!

As an aside, all our external links are loading in-window (rather than a new tab/window), can we change this and make it part of Jekyll's config?

mezis commented:
> As an aside, all our external links are loading in-window (rather than a new tab/window), can we change this and make it part of Jekyll's config?

I don't think we can, no. Unfortunately. This said blogs that open all links in a new tab are a slightly cringe-worthy experience (to me!)
